### PR TITLE
nvibrant: add module

### PR DESCRIPTION
**Pull request by mikaeladev**
### Description

based on [mikaeladev/nix-nvibrant](https://github.com/mikaeladev/nix-nvibrant) with an added option for customising dithering and support for multi-gpu setups. now that nvibrant's in nixpkgs i figured i might as well upstream the modules too.

repo: [Tremeschin/nvibrant](https://github.com/tremeschin/nvibrant)
package: [NixOS/nixpkgs](https://github.com/nixos/nixpkgs/blob/master/pkgs/by-name/nv/nvibrant/package.nix)

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -A dev --run treefmt`.

- [x] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [x] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [x] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

### CODE REVIEW

File: modules/services/nvibrant.nix [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      {
+        assertion = !(cfg.dithering == [ ] && cfg.vibrancy == [ ]);
+        message = "Either `services.nvibrant.dithering` or `services.nvibrant.vibrancy` must be set";
+      }
+    ];
+
+    systemd.user.services.nvibrant = {
+      Unit = {
+        Description = "Applies nvibrant";
+        After = [ "graphical.target" ];
```
Comment (khaneliman) [inline modules/services/nvibrant.nix]: This is a user service that needs an active graphical session and display outputs. `graphical.target` is a system target name, so ordering after it in the user manager does not make this session-bound. Please tie it to `graphical-session.target` like the other display/session services.

```suggestion
      Unit = {
        Description = "Applies nvibrant";
        After = [ "graphical-session.target" ];
        PartOf = [ "graphical-session.target" ];
      };
```

File: modules/services/nvibrant.nix [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+              mkLines (value: "${binPath} ${mkVibrancyArgs value}") cfg.vibrancy
+            );
+
+            scriptLines = concatLines (ditheringLines ++ vibrancyLines);
+          in
+          writeShellScript "apply-nvibrant" scriptLines;
+      };
+      Install = {
+        WantedBy = [ "default.target" ];
+      };
```
Comment (khaneliman) [inline modules/services/nvibrant.nix:194]: Same target issue here: enabling this under `default.target` can run it when the user manager starts, before the compositor has established the outputs.

```suggestion
      Install = {
        WantedBy = [ "graphical-session.target" ];
      };
```

File: modules/services/nvibrant.nix [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+{
+  meta.maintainers = with maintainers; [ mikaeladev ];
+
+  options.services.nvibrant = {
+    enable = mkEnableOption "nvibrant";
+
+    package = mkPackageOption pkgs "nvibrant" { };
+
+    dithering = mkOption {
+      type = with types; listOf (oneOrListOf (nullOr (either bool (enum [ "auto" ]))));
```
Comment (khaneliman) [inline modules/services/nvibrant.nix]: This type accepts mixed shapes, but `mkLines` decides single-GPU vs multi-GPU from only `head list`. For example, `[ [ true ] null ]` is accepted by the type and later fails with `expected a list but found null`; `[ true [ null ] ]` is accepted and silently renders the nested list as another disabled output. Please make the shape strict at the option boundary.

```suggestion
      type =
        let
          value = nullOr (either bool (enum [ "auto" ]));
        in
        with types; either (listOf value) (listOf (listOf value));
```

File: modules/services/nvibrant.nix [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        value is passed, nvibrant will default to false (even if a device is
+        connected at that port).
+
+        If you have multiple GPUs, you can pass a list of lists in order of
+        device.
+      '';
+    };
+
+    vibrancy = mkOption {
+      type = with types; listOf (oneOrListOf (nullOr (percentStrBetween 0 200)));
```
Comment (khaneliman) [inline modules/services/nvibrant.nix]: Same shape issue as `dithering`: this should accept either one flat GPU list or a list of GPU lists, but not a mixed list.

```suggestion
      type =
        let
          value = nullOr (percentStrBetween 0 200);
        in
        with types; either (listOf value) (listOf (listOf value));
```

**Review by khaneliman — CHANGES_REQUESTED** [review]
Two things look worth fixing before merge: the user service is tied to the wrong target for a display-dependent command, and the option type accepts mixed shapes that the generator cannot handle consistently.

File: modules/services/nvibrant.nix (hunk 3 above)
Comment (mikaeladev) [reply modules/services/nvibrant.nix]: multigpu tests fail when using this. i encountered this in initial prototyping and chalked it up to some weird edge case to figure out later on.

```nix
listOrListOfListOf = t: with types; either (listOf t) (listOf (listOf t));
type = with types; listOrListOfListOf (nullOr (either bool (enum [ "auto" ])));
```

changing the order of `listOrListOfListOf` swaps which tests fail. any ideas?

Comment (mikaeladev) [reply modules/services/nvibrant.nix]: pushing changes despite this, i'll mess around with it for now

Comment (khaneliman) [reply modules/services/nvibrant.nix]: Sorry, my earlier suggestion missed how `types.either` merges `listOf` branches. `listOf.check` is shallow, so branch order changes which valid shape fails.

Original request still stands: reject mixed flat/list entries at the option boundary. This keeps `mkLines` safe because it chooses single-GPU vs multi-GPU from the first element. I tested this locally against the nvibrant NMT tests, and it also rejects both mixed-shape examples.

```suggestion
  listOrListOfListOf =
    t:
    types.addCheck (with types; listOf (either t (listOf t))) (
      values:
      values == [ ]
      || (
        let
          isMultiGpu = isList (head values);
        in
        lib.all (value: isList value == isMultiGpu) values
      )
    );
```

Comment (mikaeladev) [reply modules/services/nvibrant.nix]: cheers!
